### PR TITLE
Additional flags for mc sql command

### PR DESCRIPTION
**Pull request by poornas**
Allow specifying input serialization, output serialization and compression type as flags to mc sql command

### CODE REVIEW

File: cmd/sql-main.go [hunk 1]
```diff
@@ -68,11 +129,220 @@ EXAMPLES:
...
+	flags := append(sqlFlags, ioFlags...)
+	for _, f := range globalFlags {
+		if f.GetName() != "json" {
+			flags = append(flags, f)
+		}
+	}
+	return flags
+}
+
+// valid CSV and JSON keys for input/output serialization
```
Comment (harshavardhana) [inline cmd/sql-main.go]: can you put this in a single var block ? 

```
var (
)
```

File: cmd/sql-main.go [hunk 2]
```diff
@@ -93,6 +363,12 @@ func sqlSelect(targetURL, expression string, encKeyDB map[string][]prefixSSEPair
 	_, e := io.Copy(os.Stdout, outputer)
 	return probe.NewError(e)
 }
```
Comment (harshavardhana) [inline cmd/sql-main.go]: needs a new line

File: cmd/sql-main.go [hunk 3]
```diff
@@ -59,6 +81,45 @@ FLAGS:
 ENVIRONMENT VARIABLES:
    MC_ENCRYPT_KEY:  list of comma delimited prefix=secret values
 
+INPUT SERIALIZATION
+	--csv-input or --json-input can be used to specify input data format. Format is specified by a string
```
Comment (harshavardhana) [inline cmd/sql-main.go]: This is really big perhaps for showing in `mc sql` we may trim it? perhaps point to the complete-guide?

File: cmd/sql-main.go (hunk 1 above)
Comment (poornas) [reply cmd/sql-main.go]: done

File: cmd/sql-main.go (hunk 3 above)
Comment (poornas) [reply cmd/sql-main.go]: done.

File: cmd/sql-main.go [hunk 4]
```diff
@@ -37,6 +43,30 @@ var (
...
+		cli.StringFlag{
+			Name:  "compression",
+			Usage: "input compression type",
+		},
+		cli.StringFlag{
+			Name:  "csv-output",
+			Usage: "csv output serialization option",
+		},
+		cli.StringFlag{
+			Name:  "csv-output-header,oh",
```
Comment (harshavardhana) [inline cmd/sql-main.go]: I don't think we need the short form `oh` it is weird to pronounce and sounds like two options crammed together.

File: cmd/sql-main.go [hunk 5]
```diff
@@ -68,11 +101,279 @@ EXAMPLES:
...
+
+// if csv-output-header is set to a comma delimited string use it, othjerwise attempt to get the header from
+// query object
+func getCSVOutputHeaders(ctx *cli.Context, url string, encKeyDB map[string][]prefixSSEPair, query string) (hdrs []string) {
+	if !ctx.IsSet("csv-output-header") {
+		return
+	}
+
+	hdrStr := ctx.String("csv-output-header")
+	if hdrStr == "" && strings.HasPrefix(strings.ToLower(query), "select *") {
```
Comment (donatello) [inline cmd/sql-main.go]: Why is the second check needed here?

Comment (poornas) [reply cmd/sql-main.go]: @donatello, csv header is being displayed if query is selecting all columns of the csv object. Hence the second check.

Comment (donatello) [reply cmd/sql-main.go]: Oh ok, maybe the check can be put in its own function and also be less strict about whitespace - maybe a regexp like `\s+select\s+\*\s+` is better.

**Comment by poornas** [comment]
> When I run this PR, and pass --csv-output-header without passing a value to it, I get the mc sql help message.

--csv-output-header is a string data type rather than a bool data type. The cli package doesnt parse the args correctly if the value is not entered.

File: cmd/sql-main.go (hunk 5 above)
Comment (poornas) [reply cmd/sql-main.go]: done

File: cmd/sql-main.go [hunk 6]
```diff
@@ -84,16 +392,34 @@ func sqlSelect(targetURL, expression string, encKeyDB map[string][]prefixSSEPair
...
 	sseKey := getSSE(targetURL, encKeyDB[alias])
-	outputer, err := targetClnt.Select(expression, sseKey)
+	outputer, err := targetClnt.Select(expression, sseKey, selOpts)
 	if err != nil {
 		return err.Trace(targetURL, expression)
 	}
 	defer outputer.Close()
-
+	if len(csvHdrs) > 0 && writeHdr {
+		fmt.Println(strings.Join(csvHdrs, ","))
```
Comment (harshavardhana) [inline cmd/sql-main.go]: Should this be removed ?

Comment (poornas) [reply cmd/sql-main.go]: no, it is writing the csv header to stdout

Comment (harshavardhana) [reply cmd/sql-main.go]: Alright can you add a comment ?

**Comment by poornas** [comment]
> Some breakage found.
> 
> ```
> $ ./mc sql -e 'select * from s3object' mys3/sqlselectapi/books.json
> mc: <ERROR> Unable to run sql Cannot format output for your query. Please check the file and query, and try again.
> ```
@donatello, fixed it

**Comment by donatello** [comment]
Found one more problem:

```
./mc sql -e 'select s.authorInfo, s.authorInfo.name from s3object s' --csv-output "rd=\n" mys3/sqlselectapi/books.json
mc: <ERROR> Unable to run sql The XML provided was not well-formed or did not validate against our published schema. Please check the service documentation and try again.
```

**Comment by donatello** [comment]
As discussed offline with @poornas the client need not send default values for csv/json format options explicitly - the server is automatically picking them. This can simplify the client.

File: go.sum [hunk 7]
```diff
@@ -87,10 +87,15 @@ github.com/minio/minio v0.0.0-20190206103305-fd4e15c11641 h1:iW6Ns9yJPCzd1v+Ajzn
...
 github.com/minio/minio-go v0.0.0-20190207005939-34f2b94c43ca/go.mod h1:vuvdOZLJuf5HmJAJrKV64MmozrSsk+or0PB5dzdfspg=
+<<<<<<< HEAD
 github.com/minio/minio-go v0.0.0-20190227180923-59af836a7e6d h1:gptD0/Hnam7h4Iq9D/33fscRpHfzOOUqUbH2nPw9HcU=
 github.com/minio/minio-go v0.0.0-20190227180923-59af836a7e6d/go.mod h1:/haSOWG8hQNx2+JOfLJ9GKp61EAmgPwRVw/Sac0NzaM=
 github.com/minio/minio-go v6.0.14+incompatible h1:fnV+GD28LeqdN6vT2XdGKW8Qe/IfjJDswNVuni6km9o=
 github.com/minio/minio-go v6.0.14+incompatible/go.mod h1:7guKYtitv8dktvNUGrhzmNlA5wrAABTQXCoesZdFQO8=
+=======
+github.com/minio/minio-go v0.0.0-20190308013636-b32976861da0 h1:yHdN+v8UkkeKXBF7evFjHE9hvdrW6WhyT/S0W07YM48=
+github.com/minio/minio-go v0.0.0-20190308013636-b32976861da0/go.mod h1:/haSOWG8hQNx2+JOfLJ9GKp61EAmgPwRVw/Sac0NzaM=
+>>>>>>> vendor in minio-go changes
```
Comment (harshavardhana) [inline go.sum]: can you run `go mod tidy` to fix these issues? @poornas

Comment (poornas) [reply go.sum]: done.

**Comment by kannappanr** [comment]
@poornas can you please take a look at Travis failure?

**Comment by poornas** [comment]
> @poornas can you please take a look at Travis failure?

golint repo is not found by travis, the make verifiers succeeds locally though with GO111MOD on
```
$ make verifiers
Installing golint
go get golang.org/x/lint/golint: git ls-remote -q https://go.googlesource.com/lint in /home/travis/gopath/pkg/mod/cache/vcs/79f2695321bc515689ad880a6109a13399c003ff78670317e7d502240c43b765: exit status 128:
	fatal: remote error: Git repository not found
```

**Pull request merged**